rules/camunda-cloud/no-loop.js

```diff
@@ -56,51 +66,112 @@ module.exports = skipInNonExecutableProcess(function() {
   };
 });
 
-function findLoop(flowElement, parentElement, visitedFlowElements = []) {
+/**
+ * @typedef {Object} GraphNode
+ * @property {ModdleElement} element the bpmn element this node represents
+ * @property {Map<ModdleElement, Array<ModdleElement>>} incoming Maps the target node with the shortest path to it
+ * @property {Map<ModdleElement, Array<ModdleElement>>} outgoing Maps the source node with the shortest path to it
+ */
 
-  // (1.1) is not a loop
-  if (!isAnyExactly(flowElement, LOOP_ELEMENT_TYPES)) {
-    return null;
-  }
+/**
+ * Simplifies the graph by removing all non-`LOOP_REQUIRED_ELEMENT_TYPES` elements and connecting incoming and outgoing nodes directly.
+ * Annotates the edges with the original path. Uses breadth-first search to find paths.
+ *
+ * @param {Array<ModdleElement>} flowElements
+ * @returns {Map<ModdleElement, GraphNode>}
+ */
+function simplifyGraph(flowElements) {
 
-  const nextFlowElements = getNextFlowElements(flowElement);
+  // Transform Array<ModdleElement> into Map<ModdleElement, GraphNode>
+  const graph = elementsToGraph(flowElements);
 
-  // (1.2) is not a loop
-  if (!nextFlowElements.length) {
-    return null;
-  }
+  breadthFirstSearch(graph, (node) => {
+    const { element, outgoing } = node;
 
-  // (2) may be a loop
-  if (!visitedFlowElements.includes(flowElement)) {
-    return nextFlowElements.reduce((error, nextFlowElement) => {
-      return error || findLoop(nextFlowElement, parentElement, [ ...visitedFlowElements, flowElement ]);
-    }, null);
-  }
+    // Remove non-required element and connect incoming and outgoing nodes directly
+    if (!isAnyExactly(element, LOOP_REQUIRED_ELEMENT_TYPES)) {
+      connectNodes(graph, node);
+    }
+
+    return [ ...outgoing.keys() ].map(key => graph.get(key));
```

Comment:
```suggestion
    return [ ...outgoing.keys() ].map(key => graph.get(key));
```

We could use `Array.from(...)` to not get into infinite loop issues here.

Reply:
What do you mean by "infinite loop"? We could use `Array.from(iterable, mapFn)`, but I do not see where the current implementation creates a loop

Reply:
If you are referring to the `Maximum call stack size exceeded` issue we had in the variable resolver, this was because `.push(...array)` and other functions will create a stack entry for every element. 
Creating new arrays using the spread operator is safe, as it is not a function call

Reply:
Yea, I referred to that instance of a problem. If we have a built-in utility, we could use it. That is what I'm saying, between the lines.

Reply:
I created a follow-up here: https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/167

As the current implementation is not vulnerable to the Max Call stack issue, I see this as a simple code-style improvements which we do not need to release right away.

Reply:
Since `outgoing` is a `Map`, `outgoing.keys()` returns an `Array`, and `[].map` creates a new `Array`, we don't need to do either `[ ...outgoing.keys() ]` or `Array.from(outgoing.keys())`. It's completely safe and IMO easier to read if we just do `outgoing.keys().map(key => graph.get(key))`.

Reply:
Oh OK. Thanks for clarification and sorry for BS

Reply:
No worries, I assumed it creates an array at first as well. It can be confusing, especially since `Object.keys` returns an array 😄